File: 90.subsets-ii.java
/*
 * @lc app=leetcode id=90 lang=java
 *
 * [90] Subsets II
 */

// @lc code=start
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        List<List<Integer>> list = new ArrayList<>();
        Arrays.sort(nums); // not necessary
        boolean[] used = new boolean[nums.length];

        // result list, templist, targetlist, start
        backtrack(list, new ArrayList<>(), nums, 0, used);
        return list;
    }

    private void backtrack(List<List<Integer>> list, List<Integer> tempList, int[] nums, int start, boolean[] used) {
        // add
        list.add(new ArrayList<>(tempList));
        // for loop start at i = start
        for (int i = start; i < nums.length; i++) {
            // add to the list
            if (used[i]) continue;

            // last branch of 122 tree, we don't include 2 because nums[1]=nums[2] = 2 but nums[1] is not visited
            // By using nums[i-1]==nums[i] && !used[i-1], we can make sure that 1b cannot be chosen before 1a
            if (i > 0 && nums[i - 1] == nums[i] && !used[i - 1]) continue;
            tempList.add(nums[i]);
            used[i] = true;
            backtrack(list, tempList, nums, i + 1, used);
            used[i] = false;
            tempList.remove(tempList.size() - 1);
        }
    }

}
// @lc code=end

File: cpp/90.subsets-ii.cpp
/*
 * @lc app=leetcode id=90 lang=cpp
 *
 * [90] Subsets II
 */

// @lc code=start

#include <algorithm>
#include <vector>
using namespace std;

// class Solution {
//    public:
//     vector<vector<int>> subsetsWithDup(vector<int>& nums) {
//         // Different from 78, now we can have duplicates
//         // we can have duplicate elements but we don't want duplicate in the result
//         vector<vector<int>> result;
//         vector<int> current;
//         // since we use nums[i] == nums[i - 1] to avoid case of duplicate, we will need sort nums
//         sort(nums.begin(), nums.end());

//         backtrack(result, current, nums, 0);
//         return result;
//     };
//     void backtrack(vector<vector<int>>& result, vector<int>& current, vector<int>& nums,
//                    int start) {
//         // base case
//         result.push_back(current);

//         // loop
//         for (int i = start; i < nums.size(); i++) {
//             if (i > start && nums[i] == nums[i - 1]) {
//                 continue;
//             }

//             current.push_back(nums[i]);
//             backtrack(result, current, nums, i + 1);
//             current.pop_back();
//         }
//         // example:
//         // [1,2,2] -> [] -> [1] -> [1,2] -> [1,2,2] -> [2] -> [2,2]
//     }
// };

#include <set>
class Solution {
   public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        set<vector<int>> result_set;
        vector<int> current;
        sort(nums.begin(), nums.end());

        int index = 0;
        backtrack(result_set, current, nums, index);

        vector<vector<int>> result_vector(result_set.begin(), result_set.end());
        return result_vector;
    }

    void backtrack(set<vector<int>>& result_set, vector<int>& current, vector<int>& nums,
                   int index) {
        if (index == nums.size()) {
            result_set.insert(current);
            return;
        }
        // Note: the first backtrack to represent the choice that exclude the index element
        // second to represent the choice of include the index element
        // Hard to manipulate duplicate, so use set
        backtrack(result_set, current, nums, index + 1);
        current.push_back(nums[index]);
        backtrack(result_set, current, nums, index + 1);
        // if current is pass by value, no need to pop back
        current.pop_back();
    }
};
// @lc code=end
